File: src/ui/components/show_mnemonic.c
// Copyright 2019 Shift Cryptosecurity AG
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <stdio.h>
#include <string.h>

#include <hardfault.h>
#include <ui/components/ui_components.h>
#include <ui/ui_util.h>

#include "show_mnemonic.h"

/**
 * Component data.
 */
typedef struct {
    void (*confirm_mnemonic)(void);
    component_t* continue_button;
} show_mnemonic_data_t;

/**
 * Collects all component functions.
 */
static const component_functions_t _component_functions = {
    .cleanup = ui_util_component_cleanup,
    .render = ui_util_component_render_subcomponents,
    .on_event = ui_util_on_event_noop,
};

/**
 * Creates a screen that allows to scroll through the mnemonic words.
 * @param[in] wordlist The mnemonic sentence, split into words.
 * @param[in] length The amount of words in the mnemonic sentence.
 * @param[in] confirm_mnemonic A callback that lets the user confirm the words he/she wrote down.
 */
component_t* show_mnemonic_create(
    const char** wordlist,
    uint8_t length,
    void (*confirm_mnemonic)(void))
{
    show_mnemonic_data_t* data = malloc(sizeof(show_mnemonic_data_t));
    if (!data) {
        Abort("Error: malloc show_mnemonic data");
    }
    memset(data, 0, sizeof(show_mnemonic_data_t));

    data->confirm_mnemonic = confirm_mnemonic;

    component_t* show_mnemonic = malloc(sizeof(component_t));
    if (!show_mnemonic) {
        Abort("Error: malloc show_mnemonic");
    }
    memset(show_mnemonic, 0, sizeof(component_t));
    show_mnemonic->f = &_component_functions;

    show_mnemonic->data = data;
    show_mnemonic->dimension.width = SCREEN_WIDTH;
    show_mnemonic->dimension.height = SCREEN_HEIGHT;

    ui_util_add_sub_component(
        show_mnemonic,
        scroll_through_all_variants_create(
            wordlist, NULL, length, true, confirm_mnemonic, show_mnemonic));

    return show_mnemonic;
}

File: src/rust/bitbox02-rust/src/hww/api/show_mnemonic.rs
// Copyright 2020 Shift Crypto AG
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use alloc::vec::Vec;

use super::Error;
use crate::pb;

use pb::response::Response;

use crate::hal::Ui;
use crate::workflow::{confirm, unlock};

/// Handle the ShowMnemonic API call. This shows the seed encoded as
/// 12/18/24 BIP39 English words. Afterwards, for each word, the user
/// is asked to pick the right word among 5 words, to check if they
/// wrote it down correctly.
pub async fn process(hal: &mut impl crate::hal::Hal) -> Result<Response, Error> {
    let mnemonic_sentence = {
        let seed = if bitbox02::memory::is_initialized() {
            unlock::unlock_keystore(hal, "Unlock device", unlock::CanCancel::Yes).await?
        } else {
            crate::keystore::copy_seed(hal)?
        };

        crate::bip39::mnemonic_from_seed(&seed)?
    };

    hal.ui()
        .confirm(&confirm::Params {
            title: "Warning",
            body: "DO NOT share your\nrecovery words with\nanyone!",
            accept_is_nextarrow: true,
            ..Default::default()
        })
        .await?;

    hal.ui()
        .confirm(&confirm::Params {
            title: "Recovery\nwords",
            body: "Please write down\nthe following words",
            accept_is_nextarrow: true,
            ..Default::default()
        })
        .await?;

    let words: Vec<&str> = mnemonic_sentence.split(' ').collect();

    hal.ui().show_and_confirm_mnemonic(&words).await?;

    bitbox02::memory::set_initialized().or(Err(Error::Memory))?;

    hal.ui().status("Backup created", true).await;
    Ok(Response::Success(pb::Success {}))
}

#[cfg(test)]
mod tests {
    use super::*;

    use alloc::boxed::Box;

    use crate::hal::testing::TestingHal;
    use crate::workflow::testing::Screen;
    use bitbox02::testing::mock_memory;
    use util::bb02_async::block_on;

    /// When not yet initialized, we show the mnemonic without a password check. This happens during
    /// wallet setup.
    #[test]
    fn test_process_uninitialized() {
        mock_memory();
        crate::keystore::encrypt_and_store_seed(
            &mut TestingHal::new(),
            hex::decode("c7940c13479b8d9a6498f4e50d5a42e0d617bc8e8ac9f2b8cecf97e94c2b035c")
                .unwrap()
                .as_slice(),
            "password",
        )
        .unwrap();

        assert!(!bitbox02::memory::is_initialized());

        let mut mock_hal = TestingHal::new();
        mock_hal.ui.set_enter_string(Box::new(|_params| {
            panic!("unexpected call to enter password")
        }));

        mock_hal.securechip.event_counter_reset();
        assert_eq!(
            block_on(process(&mut mock_hal)),
            Ok(Response::Success(pb::Success {}))
        );
        // 1 operation for one copy_seed() to get the seed to display it.
        assert_eq!(mock_hal.securechip.get_event_counter(), 1);

        assert_eq!(
            mock_hal.ui.screens,
            vec![
                Screen::Confirm {
                    title: "Warning".into(),
                    body: "DO NOT share your\nrecovery words with\nanyone!".into(),
                    longtouch: false
                },
                Screen::Confirm {
                    title: "Recovery\nwords".into(),
                    body: "Please write down\nthe following words".into(),
                    longtouch: false
                },
                Screen::ShowAndConfirmMnemonic {
                    mnemonic: "shy parrot age monkey rhythm snake mystery burden topic hello mouse script gesture tattoo demand float verify shoe recycle cool network better aspect list".into(),
                },
                Screen::Status {
                    title: "Backup created".into(),
                    success: true
                },
            ]
        );
    }
    /// When initialized, a password check is prompted before displaying the mnemonic.
    #[test]
    fn test_process_initialized() {
        mock_memory();
        crate::keystore::encrypt_and_store_seed(
            &mut TestingHal::new(),
            hex::decode("c7940c13479b8d9a6498f4e50d5a42e0d617bc8e8ac9f2b8cecf97e94c2b035c")
                .unwrap()
                .as_slice(),
            "password",
        )
        .unwrap();

        bitbox02::memory::set_initialized().unwrap();

        let mut password_entered: bool = false;

        let mut mock_hal = TestingHal::new();
        mock_hal.ui.set_enter_string(Box::new(|_params| {
            password_entered = true;
            Ok("password".into())
        }));

        mock_hal.securechip.event_counter_reset();
        assert_eq!(
            block_on(process(&mut mock_hal)),
            Ok(Response::Success(pb::Success {}))
        );
        assert_eq!(mock_hal.securechip.get_event_counter(), 5);

        assert_eq!(
            mock_hal.ui.screens,
            vec![
                Screen::Confirm {
                    title: "Warning".into(),
                    body: "DO NOT share your\nrecovery words with\nanyone!".into(),
                    longtouch: false
                },
                Screen::Confirm {
                    title: "Recovery\nwords".into(),
                    body: "Please write down\nthe following words".into(),
                    longtouch: false
                },
                Screen::ShowAndConfirmMnemonic {
                    mnemonic: "shy parrot age monkey rhythm snake mystery burden topic hello mouse script gesture tattoo demand float verify shoe recycle cool network better aspect list".into(),
                },
                Screen::Status {
                    title: "Backup created".into(),
                    success: true
                },
            ]
        );

        drop(mock_hal); // to remove mutable borrow of `password_entered`
        assert!(password_entered);
    }

    /// When initialized, a password check is prompted before displaying the mnemonic.
    /// This tests that we fail early if the wrong password is entered.
    #[test]
    fn test_process_initialized_wrong_password() {
        mock_memory();
        crate::keystore::encrypt_and_store_seed(
            &mut TestingHal::new(),
            hex::decode("c7940c13479b8d9a6498f4e50d5a42e0d617bc8e8ac9f2b8cecf97e94c2b035c")
                .unwrap()
                .as_slice(),
            "password",
        )
        .unwrap();

        bitbox02::memory::set_initialized().unwrap();

        let mut mock_hal = TestingHal::new();
        mock_hal
            .ui
            .set_enter_string(Box::new(|_params| Ok("wrong password".into())));

        mock_hal.securechip.event_counter_reset();
        assert_eq!(block_on(process(&mut mock_hal)), Err(Error::Generic));
        assert_eq!(mock_hal.securechip.get_event_counter(), 5);

        assert_eq!(
            mock_hal.ui.screens,
            vec![Screen::Status {
                title: "Wrong password\n9 tries remain".into(),
                success: false,
            }]
        );
    }
}
